crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -206,13 +206,51 @@ impl<'db> SemanticIndexBuilder<'db> {
     }
 
     fn pop_scope(&mut self) -> FileScopeId {
-        let ScopeInfo { file_scope_id, .. } =
-            self.scope_stack.pop().expect("Root scope to be present");
-        let children_end = self.scopes.next_index();
-        let scope = &mut self.scopes[file_scope_id];
-        scope.descendents = scope.descendents.start..children_end;
         self.try_node_context_stack_manager.exit_scope();
-        file_scope_id
+
+        let ScopeInfo {
+            file_scope_id: popped_scope_id,
+            ..
+        } = self
+            .scope_stack
+            .pop()
+            .expect("Root scope should be present");
+
+        let mut scopes = std::mem::take(&mut self.scopes);
+        let children_end = scopes.next_index();
+        let popped_scope = &mut scopes[popped_scope_id];
+        popped_scope.extend_descendents(children_end);
+
+        // If the scope that we just popped off is an eager scope, we need to "lock" our view of
+        // which bindings and/or declarations reach each of the uses in the scope.
+        if let Some(eager_nested_scope) = popped_scope.node().as_eager_nested_scope() {
+            for nested_symbol in self.symbol_tables[popped_scope_id].symbols() {
+                // Loop through each enclosing scope, looking for the innermost one that binds or
+                // declares the symbol.
+                for enclosing_scope_info in self.scope_stack.iter().rev() {
+                    let enclosing_scope_id = enclosing_scope_info.file_scope_id;
+                    let enclosing_symbol_table = &self.symbol_tables[enclosing_scope_id];
+                    if let Some(enclosing_symbol_id) =
+                        enclosing_symbol_table.symbol_id_by_name(nested_symbol.name())
+                    {
+                        let enclosing_symbol = enclosing_symbol_table.symbol(enclosing_symbol_id);
+                        if enclosing_symbol.is_bound() || enclosing_symbol.is_declared() {
+                            let eager_nested_scope_id = self.ast_ids[enclosing_scope_id]
```

Comment:
Another bit of logic that we handle in `infer_name_load` but I don't see here is the bit about how once we pass through an ancestor lazy scope, we are now a lazy reference, no matter what eager scope we originated in. Perhaps the only consequence of this is that we take a snapshot in some enclosing scope that we'll never use? But it still seems like it would be better to avoid doing that.